python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -528,100 +562,275 @@ def asset_key_for_input(self, input_name: str) -> AssetKey:
+
+                # running a backfill of the 2023-08-21 through 2023-08-25 partitions of this asset will log:
+                #   PartitionKeyRange(start="2023-08-20", end="2023-08-24")
+                #   PartitionKeyRange(start="2023-08-21", end="2023-08-25")
 
-    @public
-    def asset_partition_keys_for_output(self, output_name: str = "result") -> Sequence[str]:
-        """Returns a list of the partition keys for the given output."""
-        return self.asset_partitions_def_for_output(output_name).get_partition_keys_in_range(
-            self._step_execution_context.asset_partition_key_range_for_output(output_name),
-            dynamic_partitions_store=self.instance,
+        """
+        return self._step_execution_context.partition_key_range_for_asset(
+            asset, is_dependency=is_dependency
         )
 
     @public
-    def asset_partition_keys_for_input(self, input_name: str) -> Sequence[str]:
-        """Returns a list of the partition keys of the upstream asset corresponding to the
-        given input.
+    def partitions_def_for_asset(self, asset: CoercibleToAssetKey) -> PartitionsDefinition:
+        """The PartitionsDefinition of the provided asset.
+
+        Args:
+            asset (Union[AssetKey, str]): The asset to get the partition time window for
+
+        Examples:
+            .. code-block:: python
+
+                partitions_def = DailyPartitionsDefinition("2023-08-20")
+
+                @asset(
+                    partitions_def=partitions_def
+                )
+                def upstream_asset(context: AssetExecutionContext):
+                    context.log.info(context.partitions_def_for_asset("upstream_asset"))
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   DailyPartitionsDefinition("2023-08-20")
+
+                @asset(
+                    partitions_def=partitions_def,
+                )
+                def downstream_asset(context: AssetExecutionContext, upstream_asset):
+                    context.log.info(context.partitions_def_for_asset("upstream_asset"))
+                    context.log.info(context.partitions_def_for_asset("downstream_asset"))
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   DailyPartitionsDefinition("2023-08-20")
+                #   DailyPartitionsDefinition("2023-08-20")
+
         """
-        return list(
-            self._step_execution_context.asset_partitions_subset_for_input(
-                input_name
-            ).get_partition_keys()
-        )
+        return self._step_execution_context.partitions_def_for_asset(asset)
 
     @public
-    def asset_partitions_time_window_for_input(self, input_name: str = "result") -> TimeWindow:
-        """The time window for the partitions of the input asset.
+    def partition_keys_for_asset(
+        self, asset: CoercibleToAssetKey, is_dependency: bool = False
```

Comment:
`is_dependency` seems like a rough name for this

Do we have a generalized way to describe this concept? Asset keys + partition mappings end up being a different sort of DAG, which you can visualize as a third dimension in the asset graph (imagine each partition being an entry in the z axis). How do we describe this elsewhere?

Reply:
+1 that the name sucks.  

I'm not aware of any generalized way to describe this kind of thing. Maybe @clairelin135  has an idea from when the partition mapping stuff was first written?